features/device-config/monitor-adaptor/src/main/java/org/opennms/features/deviceconfig/monitor/adaptor/DeviceConfigMonitorAdaptor.java

```diff
@@ -0,0 +1,130 @@
+import org.opennms.features.deviceconfig.persistence.api.DeviceConfig;
+import org.opennms.features.deviceconfig.persistence.api.DeviceConfigDao;
+import org.opennms.netmgt.dao.api.IpInterfaceDao;
+import org.opennms.netmgt.model.OnmsIpInterface;
+import org.opennms.netmgt.poller.MonitoredService;
+import org.opennms.netmgt.poller.PollStatus;
+import org.opennms.netmgt.poller.ServiceMonitorAdaptor;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import java.nio.charset.Charset;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.Map;
+import java.util.Optional;
+
+public class DeviceConfigMonitorAdaptor implements ServiceMonitorAdaptor {
+
+    private static final String DEVICE_CONFIG_MONITOR = "DeviceConfig";
+
+    @Autowired
+    private DeviceConfigDao deviceConfigDao;
+
+    @Autowired
+    private IpInterfaceDao ipInterfaceDao;
+
+    @Override
+    public PollStatus handlePollResult(MonitoredService svc, Map<String, Object> parameters, PollStatus status) {
+
+        if (!svc.getSvcName().equals(DEVICE_CONFIG_MONITOR)) {
+            return status;
+        }
+        // Retrieve interface
+        final OnmsIpInterface ipInterface = ipInterfaceDao.findByNodeIdAndIpAddress(svc.getNodeId(), svc.getIpAddr());
+        String encodingAttribute = getObjectAsString(parameters.get("encoding"));
+        String configTypeAttribute = getObjectAsString(parameters.get("config-type"));
+        String encoding = !Strings.isNullOrEmpty(encodingAttribute) ? encodingAttribute : Charset.defaultCharset().name();
+        ConfigType configType = !Strings.isNullOrEmpty(configTypeAttribute) ? ConfigType.valueOf(configTypeAttribute) : ConfigType.Default;
+
+        byte[] deviceConfigBytes = status.getDeviceConfig();
+        // Handle config retrieval failure.
+        if (deviceConfigBytes == null) {
+            Date currentTime = new Date();
+            Optional<DeviceConfig> configOptional = deviceConfigDao.getLatestConfigForInterface(ipInterface, configType);
+            DeviceConfig deviceConfig;
+            if (configOptional.isPresent() && configOptional.get().getLastFailed() != null) {
+                deviceConfig = configOptional.get();
+            } else {
+                deviceConfig = new DeviceConfig();
+                deviceConfig.setIpInterface(ipInterface);
+                deviceConfig.setConfigType(configType);
+                deviceConfig.setCreatedTime(currentTime);
+                deviceConfig.setEncoding(encoding);
+            }
+            deviceConfig.setFailureReason(status.getReason());
+            deviceConfig.setLastFailed(currentTime);
+            deviceConfigDao.saveOrUpdate(deviceConfig);
+            return status;
+        }
+
+        // Fetch last known config for the interface
+        Optional<DeviceConfig> deviceConfigOptional = deviceConfigDao.getLatestSucceededConfigForInterface(ipInterface, configType);
```

Comment:
I think we should not retrieve the latest succeeded but just the latest device config. If so, then this line could be moved up in the method because the last device config is needed both, in the failure case and the success case.

In case the last device config has a failure stored, that failure is cleared.